### PR TITLE
OAuthCard support for streaming connections

### PR DESCRIPTION
**Pull request by Jeffders**
Added support to perform GetToken requests for a period of time if an OAuthCard is sent via a streaming connection.

This includes a unit test that validates that the OAuthCard is detected, that polling is kicked off, and that an EventActivity is eventually returned to the bot.

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
+﻿using System;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TokenResolver.cs]: // Copyright

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 2]
```diff
@@ -321,12 +329,15 @@ private async Task SendOAuthCardAsync(ITurnContext turnContext, IMessageActivity
...
                                 Text = _settings.Text,
                                 Type = ActionTypes.Signin,
+                                Value = signInLink,
                             },
                         },
                     },
                 });
             }
 
+            turnContext?.TurnState?.Add<object>(TurnStateConstants.OAuthLoginTimeoutKey, _settings.Timeout);
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: This line seems very complex, with nulls and variations. Please add a detailed comment.

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 3]
```diff
@@ -431,5 +442,10 @@ private bool ChannelSupportsOAuthCard(string channelId)
 
             return true;
         }
+
+        private bool RequiresLink(ITurnContext turnContext)
+        {
+            return !turnContext.Activity.ServiceUrl.Contains("http");
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: I don't understand this. Can you add in a detailed comment explaining the logic here? If nothing else, I don't understand why we would ever allow this over http vs https.

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 4]
```diff
@@ -402,15 +403,24 @@ public override async Task<ResourceResponse[]> SendActivitiesAsync(ITurnContext
...
-                else if (!string.IsNullOrWhiteSpace(activity.ReplyToId))
-                {
-                    var connectorClient = turnContext.TurnState.Get<IConnectorClient>();
-                    response = await connectorClient.Conversations.ReplyToActivityAsync(activity, cancellationToken).ConfigureAwait(false);
-                }
                 else
                 {
-                    var connectorClient = turnContext.TurnState.Get<IConnectorClient>();
-                    response = await connectorClient.Conversations.SendToConversationAsync(activity, cancellationToken).ConfigureAwait(false);
+                    if (!activity.ServiceUrl.StartsWith("http"))
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: I'm skeptical of this code. It seems really fragile. 

Can you add in a more detailed explication of what's going on?

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder
+{
+    internal class TokenResolver
+    {
+        private const int PollingInterval = 1000;   // Poll for token every 1 second.
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TokenResolver.cs]: PollingIntervalMs

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                        // Poll as a background task
+                        Task.Run(() => PollForTokenAsync(adapter, turnContext, activity, oauthCard.ConnectionName, cancellationToken));
+                    }
+                }
+            }
+        }
+
+        private static async Task PollForTokenAsync(BotFrameworkAdapter adapter, ITurnContext turnContext, Activity activity, string connectionName, CancellationToken cancellationToken)
+        {
+            TokenResponse tokenResponse = null;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TokenResolver.cs]: Exception leakage here, based on the Task.run, doesn't seem correct.

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            bool shouldEndPolling = false;
+            int pollingTimeout = TurnStateConstants.OAuthLoginTimeoutMsValue;
+            int pollingRequestsInterval = PollingInterval;
+            var loginTimeout = turnContext?.TurnState?.Get<object>(TurnStateConstants.OAuthLoginTimeoutKey) as int?;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TokenResolver.cs]: I took a poll for this line of code and the feedback wasn't kind. 

Honestly, this is really fragile and bug prone.

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+            // Override login timeout with value set from the OAuthPrompt or by the developer
+            if (loginTimeout != null)
+            {
+                pollingTimeout = (int)loginTimeout;
+            }
+
+            if (activity.From == null || string.IsNullOrWhiteSpace(activity.From.Id))
+            {
+                throw new ArgumentNullException("TokenResolver.PollForTokenAsync: missing Activity from or from.id");
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TokenResolver.cs]: What happens to exceptions thrown here? This method is wrapped in a Task.Run() with no exception handling. I don't know if this is the behavior we want. 

https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/exception-handling-task-parallel-library

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                throw new ArgumentNullException(nameof(connectionName));
+            }
+
+            var stopwatch = Stopwatch.StartNew();
+
+            while (stopwatch.Elapsed < TimeSpan.FromMilliseconds(pollingTimeout) && !shouldEndPolling)
+            {
+                tokenResponse = await adapter.GetUserTokenAsync(turnContext, connectionName, null, cancellationToken).ConfigureAwait(false);
+
+                if (tokenResponse != null)
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TokenResolver.cs]: The null checking through here makes this code really complex. IS there a way to re-write this w/o all of the null programming?

File: libraries/Microsoft.Bot.Builder/TurnStateConstants.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TurnStateConstants.cs]: // (C)

File: libraries/Microsoft.Bot.Builder/TurnStateConstants.cs [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        /// <summary>
+        /// TurnState key for the OAuth login timeout.
+        /// </summary>
+        public const string OAuthLoginTimeoutKey = "LoginTimeout";
+
+        /// <summary>
+        /// Default amount of time an OAuthCard will remain active.
+        /// </summary>
+        public const int OAuthLoginTimeoutMsValue = 900000;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TurnStateConstants.cs]: 900,000 == 15 minutes. Any reason this value was chosen? Please add a comment.

File: libraries/Microsoft.Bot.Builder/TurnStateConstants.cs [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+        /// <summary>
+        /// Name of the token polling settings key.
+        /// </summary>
+        public const string TokenPollingSettingsKey = "tokenPollingSettings";
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TurnStateConstants.cs]: Why is this lower camel case, and the OAuthLoginTimeoutKey is upper camel case?

File: libraries/Microsoft.Bot.Schema/TokenPollingSettings.cs [hunk 13]
```diff
@@ -0,0 +1,28 @@
+﻿using Newtonsoft.Json;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Schema/TokenPollingSettings.cs]: // (C)

File: tests/Microsoft.Bot.Builder.Tests/MemoryConnectorClient.cs [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+using Microsoft.Rest;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+using Moq.Protected;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Tests
+{
+    public class MemoryConnectorClient : IConnectorClient
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.Tests/MemoryConnectorClient.cs]: Why is this here?

File: tests/Microsoft.Bot.Builder.Tests/MemoryConversations.cs [hunk 15]
```diff
@@ -0,0 +1,90 @@
+﻿using System;
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.Tests/MemoryConversations.cs]: I don't understand why this file is included.

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 16]
```diff
@@ -58,7 +58,7 @@ public class OAuthPrompt : Dialog
         private const string PersistedExpires = "expires";
 
         // Default prompt timeout of 15 minutes (in ms)
-        private const int DefaultPromptTimeout = 900000;
+        private const int DefaultPromptTimeout = TurnStateConstants.OAuthLoginTimeoutMsValue;
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: Making timeouts of type TimeSpan makes them more readable btw, i.e. TimeSpan.FromMinutes(15) is way more readable than 900000 milliseconds (i.e. I have to do math to understand how long it is)

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs (hunk 3 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: +1, is the logic here that if we are using websockets or pipes this will return true, and for http / https return false? If that is the case, this logic should be documented. Documentation could even start at the class design level, i.e. you could have a small abstraction that given an activity, determines whether it is from Pipes, Sockets, or Http. But yes, at least add comments. What are the possible values? what are the scenarios behind each value? why do some scenarios require  a link and others not? future maintainers of this code need to understand this decision fully in the future.

Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: Bug alert: Note that you are using Contains instead of StartsWith which is a bug. StartsWith catch things like "http://myservice.com", but since you used contains, I could have something like "ws://myhttptest" and pass as http, despite being websockets.

Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: bug alert 2: how should strng case affect this path? is service url guaranteed to be lowercase? Otherwise, how do you want to compare strings?

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: urls tend to be case insensitive

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 4 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: This should be centralized - you already ahve the same code in the RequiresLink method. Move this logic to a small class that takes an activity and determines the type of connection, and has meaningful documentation and names that allow the code to be more easily understandable by future maintainers. Thanks!

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 17]
```diff
@@ -402,15 +403,24 @@ public override async Task<ResourceResponse[]> SendActivitiesAsync(ITurnContext
...
+                    {
+                        // Streaming connection post-processing
+                        TokenResolver.CheckForOAuthCards(this, turnContext, activity, cancellationToken);
+                    }
+
+                    if (!string.IsNullOrWhiteSpace(activity.ReplyToId))
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Add comment: why? what does it mean that the activity has a non-empty ReplyToId?

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs (hunk 5 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder/TokenResolver.cs]: I suggest using TimeSpan, more readable. Most methods receive timeouts as TimeSpan or int, so we can use that to our advantage as well. Note below how you don't need the comment because the code itself is self-explanatory

` private const TimeSpan PollingInterval = TimeSpan.FromSeconds(1);`

Then you can pass that TimeSpan to whatever time out, or you can get the milliseconds using the TotalMilliseconds property (IIRC)

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+using System.Security.Principal;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder
+{
+    internal class TokenResolver
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder/TokenResolver.cs]: If all the methods are static, mark the class static

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+        {
+            if (activity?.Attachments != null)
+            {
+                foreach (var attachment in activity?.Attachments?.Where(a => a.ContentType == OAuthCard.ContentType))
+                {
+                    OAuthCard oauthCard = attachment.Content as OAuthCard;
+                    if (oauthCard != null && !string.IsNullOrEmpty(oauthCard.ConnectionName))
+                    {
+                        // Poll as a background task
+                        Task.Run(() => PollForTokenAsync(adapter, turnContext, activity, oauthCard.ConnectionName, cancellationToken));
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder/TokenResolver.cs]: Have you evaluated the advantages of using this vs `HostingEnvironment.QueueBackgroundWorkItem`? SEe here for more details: https://www.hanselman.com/blog/HowToRunBackgroundTasksInASPNET.aspx.

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs (hunk 7 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder/TokenResolver.cs]: And here you get a int? that below you cast to an int. this seems a bit overcomplicated.

File: libraries/Microsoft.Bot.Builder/TurnStateConstants.cs (hunk 11 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder/TurnStateConstants.cs]: Also, I suggest `public const TimeSpan OAuthLoginTimeout = TimeSpan.FromMinutes(15);`

File: tests/Microsoft.Bot.Builder.Tests/BotFrameworkAdapterOAuthTests.cs [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+using Moq.Protected;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Tests
+{
+    [TestClass]
+    public class BotFrameworkAdapterOAuthTests
+    {
+        [TestMethod]
+        public async Task BotFrameworkAdapterFindsOAuthCards()
```
Comment (carlosscastro) [inline tests/Microsoft.Bot.Builder.Tests/BotFrameworkAdapterOAuthTests.cs]: A single test for all the possibilities that your code is introducing seems insufficient for me.

For example, you had lines that had 3 question marks, that means that 3 things could be null or none be null. That is 4 cases. 

OTher relevant tests:

-Polling until timeout: do things end correctly?
-Polling and getting the token first try
-Polling and getting the token second try 

Can this be tested at all? Seems more like functional tests but if they could be done that would be great. We can add secrets in the build system if you need any, let me know, happy to help here.

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs (hunk 3 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: This is documented in OBI, which is not public, so I will add a comment.

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 4 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: This is all well and good, and I'll do that...but because Prompts and Adapters live in different assemblies I will now have to make this method public, which means when the rules change in the future this method will change.

File: libraries/Microsoft.Bot.Builder/TurnStateConstants.cs (hunk 11 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/TurnStateConstants.cs]: I didn't chose it, Steve did (see, I removed where he put it only in the OAuthPrompt and moved it more centrally). It's just the amount of time the card stays active and has been there forever, but I'll add a comment now and change it to a TimeSpan

File: tests/Microsoft.Bot.Builder.Tests/MemoryConnectorClient.cs (hunk 14 above)
Comment (Jeffders) [reply tests/Microsoft.Bot.Builder.Tests/MemoryConnectorClient.cs]: I use it in my tests...what do you mean? 

It's used to capture the calls that would be made to the ConnectorClient so I can run through tests Functionally without having to make any calls.

File: tests/Microsoft.Bot.Builder.Tests/MemoryConversations.cs (hunk 15 above)
Comment (Jeffders) [reply tests/Microsoft.Bot.Builder.Tests/MemoryConversations.cs]: It's used in my tests.

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs (hunk 16 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: Yes, I had removed this from here, and where I put it I changed it to a TimeStamp.

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs (hunk 3 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: I changed it to StartsWith. Thank you.

Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: Good catch.

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 17 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Sorry, if you look, I didn't change this, I simply indented it. I believe from protocol that ReplyToId is not required, but if present controls whether this is a reply or just sending to a conversation. Bu I will add that comment.

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs (hunk 19 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/TokenResolver.cs]: Well, as this is doing a small amount of a work at an interval, the blog says: 
"If your ASP.NET application just needs one background task to runs an a basic scheduled interval, than perhaps you just need the basics of WebBackgrounder.", which is Task.Run. 

Also, when I reached out to the ASP.NET team a few weeks ago to talk to them about thread pools, they said with ASP.NET Core there isn't a a separate thread pool anymore and that Task.Run was functionally equivalent to more complex mechanisms (except when you really want your own task scheduler). This could be Core or NetFx I guess, but still it's just running at an interval.

That's a good article though :-)

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs (hunk 9 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/TokenResolver.cs]: This method on the adapter can return null so we have to deal with that case. It means "the token isn't there yet" and is normal.

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs (hunk 5 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/TokenResolver.cs]: Went with Carlos's suggestion

File: tests/Microsoft.Bot.Builder.Tests/MemoryConnectorClient.cs (hunk 14 above)
Comment (Jeffders) [reply tests/Microsoft.Bot.Builder.Tests/MemoryConnectorClient.cs]: Tests!

File: tests/Microsoft.Bot.Builder.Tests/MemoryConversations.cs (hunk 15 above)
Comment (Jeffders) [reply tests/Microsoft.Bot.Builder.Tests/MemoryConversations.cs]: Tests...look at my tests.

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs (hunk 6 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/TokenResolver.cs]: I added a continuation step.

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs (hunk 8 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/TokenResolver.cs]: They will now get caught by the continuation.

File: tests/Microsoft.Bot.Builder.Tests/BotFrameworkAdapterOAuthTests.cs (hunk 20 above)
Comment (Jeffders) [reply tests/Microsoft.Bot.Builder.Tests/BotFrameworkAdapterOAuthTests.cs]: I added more test cases. It can be tested with all the Mocks I made for the ConnectorClient and OAuthClient.

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+{
+    internal static class TokenResolver
+    {
+        private static readonly TimeSpan PollingInterval = TimeSpan.FromSeconds(1);   // Poll for token every 1 second.
+
+        public static void CheckForOAuthCards(BotFrameworkAdapter adapter, ILogger logger, ITurnContext turnContext, Activity activity, CancellationToken cancellationToken)
+        {
+            if (activity?.Attachments != null)
+            {
+                foreach (var attachment in activity?.Attachments?.Where(a => a.ContentType == OAuthCard.ContentType))
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TokenResolver.cs]: You just checked above for activity != null & attachments != null. 

Null checks here are not adding anything.

File: libraries/Microsoft.Bot.Builder/TokenResolver.cs [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+
+namespace Microsoft.Bot.Builder
+{
+    internal static class TokenResolver
+    {
+        private static readonly TimeSpan PollingInterval = TimeSpan.FromSeconds(1);   // Poll for token every 1 second.
+
+        public static void CheckForOAuthCards(BotFrameworkAdapter adapter, ILogger logger, ITurnContext turnContext, Activity activity, CancellationToken cancellationToken)
+        {
+            if (activity?.Attachments != null)
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/TokenResolver.cs]: If (activity == null) 
{ 
  return;
}

if (activty.achments == null)
{
  return
}

The method will be longer, but simpler.

File: libraries/Microsoft.Bot.Schema/ActivityEx.cs [hunk 23]
```diff
@@ -507,6 +507,19 @@ public Activity ApplyConversationReference(ConversationReference reference, bool
             return this;
         }
 
+        /// <summary>
+        /// Determine if the Activity was sent via an Http/Https connection or Streaming
+        /// This can be determined by looking at the ServiceUrl property:
+        /// (1) All channels that send messages via http/https are not streaming
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Schema/ActivityEx.cs]: Do we allow http channels? Other than localhost, I don't think we do.

File: libraries/Microsoft.Bot.Schema/ActivityEx.cs [hunk 24]
```diff
@@ -507,6 +507,19 @@ public Activity ApplyConversationReference(ConversationReference reference, bool
...
+        /// (2) Channels that send messages via streaming have a ServiceUrl that does not begin with http/https
+        /// </summary>
+        /// <returns>True if the Activity originated from a streaming connection.</returns>
+        public bool IsFromStreamingConnection()
+        {
+            var isHttp = ServiceUrl?.StartsWith("http", StringComparison.InvariantCultureIgnoreCase);
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Schema/ActivityEx.cs]: Use the URI class for this. The relevant property is "scheme". 
https://docs.microsoft.com/en-us/dotnet/api/system.uri.scheme?view=netframework-4.8

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)

File: libraries/Microsoft.Bot.Schema/ActivityEx.cs (hunk 23 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Schema/ActivityEx.cs]: Most channels reply with https for a service url. Local will be http. Streaming won't be either of those.

File: libraries/Microsoft.Bot.Schema/ActivityEx.cs (hunk 24 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Schema/ActivityEx.cs]: The SchemaUrl format does not strictly adhere to URI format so I cannot.

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)
